README.md
Source files for this blogdown site
===================================


The design of this blog has been inspired by the blog of https://robjhyndman.com. The theme was originally based on  [hugo-finite](https://github.com/lambdafu/hugo-finite). 
The theme `hyndman` can be found in this blog under the folder `themes`.


### Licenses of third-party components

* hugo-finite: MIT License
* Highlight.js: MIT License
* MathJax: Apache License 2.0
* Foundation 6.2.3 by Zurb: MIT License
* Font Awesome 4.6.3: SIL OFL 1.1 (font), MIT License (CSS)
* Academicons 1.7.0: SIL OFL 1.1 (font), MIT License (CSS)
* Google fonts: OFL (SIL Open Font License)



NEWS.md
# 0.1.14
* 24 Jun 2020
* fix summary of post showing content as well
* realign code under `layouts/index.html`
* add div class for post summary
* remove remnant div at the end of index.html
* add More button to posts

# 0.1.13
* 24 Jun 2020
* Fix empty summaries. Change old ``.Data.Pages.ByDate` to `.Site.RegularPages.ByDate` in page /hyndman/layouts/index.html
* Rebuild in Linux machine under R-3.6.3
* Improve Makefile. Enable rsync and clean rules


# 0.1.12
* 01 Feb 2019
* Recreate symbolic link to witsml because it was giving error. Let's not use symbolic link clone anymore.
* add note on the version of WITSML being used.

# 0.1.11
* 30 Jan 2019
* new article Exploring drilling data from the Volve dataset with WITSML and R.
* added Volve drilling datasets for three wells. 
* Used symbolic link for the witsml folder presence under content/publication
* new folder inst/rawdata under /static

# 0.1.10.9000
* change slideshare link to shortcode
* add speakerdeck slide to 'Best Practices for the Construction of Well Models"
* cannot add speakerdeck link because there is no key or id link
* add cover to Evolution of data science, machine learning and artificial intelligence
* add date to Latest Articles/Papers

# 0.1.10
* add posts for 2018 (nov, dec), 2017 (dec-aug)
* modify publications to show 
    * badge and link for SlideShare
    * badge, link and icon for GitHub
    * complete date format dd-mmm-yyyy
    * images with link to same image


# 0.1.9
* new posts. new images
* first post with links to GitHub
* adding images on top of the article instead of referring to `image:` parameter in the YAML header.


# 0.1.8
* test petro.One with post on gas lift
* add new posts, new publications
* online since Jan 02
* add Python code inside Rmarkdown notebook
* allow blog in two forms: blog.example.com and example.com/blog
* modify date format in publications
* change from robhyndman to hyndman css


# 0.1.7
* Not online yet
* Add few posts
* change site to oilgainsanalytics.com
* add handle for Twitter
* add icons. change name of icons
* change folder `talks/` to `portfolio/`
* modify 404 page

static/Rfiles/ROC.R
######################################################
### FUNCTIONS FOR KERNEL ESTIMATION OF ROC CURVES  ###
### Last updated by RJH. 12 May 2011               ###
### Version of functions for public release        ###
######################################################

#################################
### ROC CALCULATION FROM CDFS ###
#################################

# Return ROC function based on Fx and Fy inputs
# Computed at x
ROC <- function(x,Fx,y,Fy,nx=200)
{
    xx <- seq(min(x,y),max(x,y),l=nx)
    newFx <- approx(x,Fx,xout=xx,yleft=0,yright=1)
    newFy <- approx(y,Fy,xout=xx,yleft=0,yright=1)
    return(list(x=c(0,rev(1-newFx$y),1),y=c(0,rev(1-newFy$y),1)))
}

##################################################
### ROC ESTIMATION using empirical cdf         ###
##################################################

# Step function estimate of ROC curve based on empirical cdfs

ROC.rawest <- function(x,y)
{
    tmp <- reduce(x,y)
    tmpx <- ecdf(tmp$x)
    tmpy <- ecdf(tmp$y)

    xandy <- sort(unique(c(tmp$x,tmp$y)))
    xx <- knots(tmpx)
    xy <- tmpx(xx)
    yx <- knots(tmpy)
    yy <- tmpy(yx)
    Fxx <- approx(xx,xy,xandy,yleft=0,yright=1,method="constant",f=1)$y
    Fyy <- approx(yx,yy,xandy,yleft=0,yright=1,method="constant",f=1)$y
    return(list(x=c(0,rev(1-Fxx),1),y=c(0,rev(1-Fyy),1)))
}

###############################################
### ROC ESTIMATION using kernel cdf         ###
###############################################

# Kernel estimate of ROC curve with bandwidths hx and hy
# If bandwidths missing, uses method from ROC.h function
# See bandwidths.R for details.

ROC.est <- function(x,y,hx,hy,conf=95,intervals=TRUE,...)
{
    if(missing(hx) | missing(hy))
    {
        hopt <- ROC.h(x,y,...)
        hx <- hopt$hx
        hy <- hopt$hy
    }
    lo <- min(min(x)-5*hx,min(y)-5*hy)
    hi <- max(max(x)+5*hx,max(y)+5*hy)
    xgrid <- seq(lo,hi,l=200)
    Fx <- cdf.est(x,hx,xgrid)
    Fy <- cdf.est(y,hy,xgrid)

    p <- c(0,rev(1-Fx$y),1)
    Rhat <- c(0,rev(1-Fy$y),1)

    if(intervals)
    {
        tmp <- ASMCONF.est(x,y,alpha=1-conf/100)
        sig2 <- approx(tmp$p,tmp$sigma2,xout=p)$y
        z <- abs(qnorm(0.5-0.005*conf))
        lo <- pmax(Rhat - z*sig2, 0)
        hi <- pmin(Rhat + z*sig2, 1)
        return(list(x=p,y=Rhat,hx=hx,hy=hy,lo=lo, hi=hi, s2=sig2))
    }
    else
        return(list(x=p,y=Rhat,hx=hx,hy=hy))
}


###############################################
### Utility functions for kernel estimation ###
###############################################

# Evaluate K on grid
make.K <- function(x,h,xgrid,ngrid=200,lo,hi)
{
    # Construct grid if necessary
    if(missing(xgrid))
    {
        if(missing(lo))
            lo <- min(x)-3*h
        if(missing(hi))
            hi <- max(x)+3*h
        xgrid <- seq(lo,hi,l=ngrid)
    }
    # Setup necessary variables
    n <- length(x)
    ngrid <- length(xgrid)
    Kmat1 <- Kmat2 <- Kmat3 <- matrix(NA,ngrid,n)
    for(i in 1:ngrid)
    {
        xx <- (xgrid[i]-x)/h
        Kmat1[i,] <- pnorm(xx)
        Kmat2[i,] <- dnorm(xx)
        Kmat3[i,] <- -xx*Kmat2[i,]
    }
    # Return results
    return(list(x=xgrid,K1=Kmat1,K2=Kmat2,K3=Kmat3))
}

cdf.est <- function(x,h=h.lloyd2(x),xgrid)
{
    if(missing(xgrid))
        xgrid <- seq(min(x)-3*h,max(x)+3*h,l=200)
    n <- length(xgrid)
    Fx <- cumsum(density(x,bw=h,from=xgrid[1],to=xgrid[n],n=n)$y)
    return(list(x=xgrid,y=Fx/Fx[n]))
}

pdf.est <- function(x,h=bw.SJ(x),xgrid)
{
    if(missing(xgrid))
        xgrid <- seq(min(x)-3*h,max(x)+3*h,l=200)
    n <- length(xgrid)
    return(list(x=xgrid,y=density(x,bw=h,from=xgrid[1],to=xgrid[n],n=n)$y))
}

#  FUNCTION reduces (x,y) values by removing repeated x values.
#  Points with smaller y-value are retained.
#  REQUIRED ARUMENTS:",
#      x - vector of x-coordinates of function
#      y - vector of y -coordinates of function

reduce <- function (x = stop("Argument is missing"), y = stop("Argument is missing"))
{
    Y <- y[order(x)]
    X <- sort(x)
    tmp <- match(unique(X), X)
    X <- X[tmp]
    Y <- Y[tmp]
    return(list(x=X,y=Y))
}


#################################
### BANDWIDTH SELECTION       ###
### FOR KERNEL ROC ESTIMATION ###
#################################

# Main bandwidth selection function
# Method:
#     norm: assumes normal densities
#     lloyd: uses Lloyd's plug-in estimator for cdfs
#     plugin: uses plug-in method of Hall and Hyndman
#     mix: uses recommended method of Hall and Hyndman

ROC.h <- function(x,y,method="mix",...)
{
    if(method=="norm")
        return(h.norm(x,y))
    else if(method=="lloyd")
        return(list(hx=h.lloyd(x),hy=h.lloyd(y)))
    else if(method=="plugin")
        return(plugin(x,y,...))
    else if(method=="mix")
    {
        hn <- h.norm(x,y)
        hp <- plugin(x,y,...)
        return(list(hx=min(hn$hx,hp$hx),hy=min(hn$hy,hp$hy)))
    }
    else
        stop("Unknown method")
}

## NORMAL REFERENCE BANDWIDTH SELECTION

# Compute optimal h for each variable assuming normal densities
# with unknown means and variances
h.norm <- function(x,y)
{
    mx <- mean(x)
    my <- mean(y)
    sx <- sqrt(var(x))
    sy <- sqrt(var(y))
    nx <- length(x)
    ny <- length(y)
    h.norm2(mx,sx,my,sy,nx,ny)
}

# Compute optimal h for each variable assuming normal densities
# with known means and variances
h.norm2 <- function(mx,sx,my,sy,nx,ny)
{
    hx <- hx.norm(mx,sx,my,sy,nx)
    hy <- hx.norm(my,sy,mx,sx,ny)
    return(list(hx=hx,hy=hy))
}

# Optimal h for x variable assuming normal densities
# with known means and variances
hx.norm <- function(mx,sx,my,sy,n)
{
    kappa <- 0.5641896
    #kappa2 <- 1
    sum1 <- sx^2+sy^2
    sum2 <- sum1+sx^2
    sum3 <- (mx-my)^2
    f1 <- 4*sqrt(pi)*kappa * sx^3*sum1^2.5 / (sqrt(sum2)*(sy^4+sy^2*sx^2 + 2*sx^2*sum3))
    c3 <-f1*exp(sum3*sx^2/(sum1*sum2))
    return((c3/n)^(1/3))
}


## LLOYD PLUG-IN RULE FOR kernel estimate of CDF

h.lloyd <- function(x)
{
    n <- length(x)
    if(n < 2)
        stop("Insufficient data")
    s <- IQR(x) / 1.349
    mu <- mean(x)
    psi6 <- -0.5289/s^7
    h4 <- (-6/psi6/n)^(1/7)
    xx <- seq(min(x)-s,max(x)+s,l=512)
    pdf4deriv <- denest(x,deriv=4,h=h4,xgrid=xx)
    psi4 <- mean(approx(xx,pdf4deriv$y,xout=x)$y,na.rm=T)
    h2 <- (0.7979/psi4/n)^(1/5)
    pdf2deriv <- denest(x,deriv=2,h=h2,xgrid=xx)
    psi2 <- mean(approx(xx,pdf2deriv$y,xout=x)$y,na.rm=T)
    h <- (-0.5641896/psi2/n)^(1/3)
    if(h < 0)
        return(h.lloyd2(x))
    else
        return(h)
}

# Normal reference rule for kernel estimate of CDF
h.lloyd2 <- function(x)
{
    n <- length(x)
    s <- IQR(x) / 1.349
    return(1.583*s/n^(1/3))
}

# Estimate density and derivatives
denest <- function(x,deriv=0,h,xgrid,ngrid=200,lo,hi)
{
    # Construct grid if necessary
    if(missing(xgrid))
    {
        if(missing(lo))
            lo <- min(x)-3*h
        if(missing(hi))
            hi <- max(x)+3*h
        xgrid <- seq(lo,hi,l=ngrid)
    }

    # Setup necessary variables
    n <- length(x)
    xmat <- sweep(matrix(xgrid, nrow=length(xgrid), ncol=n), 2, x)/h
    if(deriv==-1)
        return(cdf.est(x,h,xgrid))
    else if(deriv==0)
        return(pdf.est(x,h,xgrid))
    else if(deriv==1)
        Kmat <- -xmat*dnorm(xmat)
    else if(deriv==2)
        Kmat <- dnorm(xmat)*(xmat^2-1)
    else if(deriv==3)
        Kmat <- dnorm(xmat)*xmat*(3-xmat^2)
    else if(deriv==4)
        Kmat <- dnorm(xmat)*(3-6*xmat^2+xmat^4)
    else
        stop("Not implemented")
    # Return results
    return(list(x=xgrid,y=apply(Kmat,1,sum)/(n*h^(deriv+1))))
}


## Utility functions for computing numerical derivatives
# Assume x is sorted in ascending order

# 2nd derivatives
deriv2 <- function(y,x)
{
    n <- length(x)
    del <- x[2]-x[1]
    c(NA,y[3:n]-2*y[2:(n-1)]+y[1:(n-2)],NA)/(del^2)
}

# 4th derivatives
deriv4 <- function(y,x)
{
    deriv2(deriv2(y,x),x)
}

### PLUG IN METHOD OF HALL AND HYNDMAN

# PLUG IN SELECTOR
pluginbands <- function(x,y)
{
    m <- length(x)
    n <- length(y)
    mx <- mean(x)
    my <- mean(y)
    sx <- sqrt(var(x))
    sy <- sqrt(var(y))
    rmin <- min(mx-5*sx,my-5*sy)
    rmax <- max(mx+5*sx,my+5*sy)

    R <- n/m
    lambda1 <- 0.5/sqrt(pi)
    lambda2 <- 0.5/sqrt(2*pi)
    kappa2 <- 1
    I1 <- abs(integrate(intf9,rmin,rmax,mx=my,sx=sy,my=mx,sy=sx)$value)
    I2 <- abs(integrate(intf3,rmin,rmax,mx=my,sx=sy,my=mx,sy=sx)$value)
    I3 <- abs(integrate(intf2,rmin,rmax,mx=my,sx=sy,my=mx,sy=sx)$value)
    H <- n^(-0.4)*((2*R*lambda1*I1 + lambda2*I2)/(2*kappa2^2*I3^2))^(0.2)

    I4 <- abs(integrate(intf5,rmin,rmax,mx=mx,sx=sx,my=my,sy=sy)$value)
    I5 <- abs(integrate(intf6,rmin,rmax,mx=mx,sx=sx,my=my,sy=sy)$value)
    J <- 64*abs(integrate(intf10,rmin,rmax,mx=mx,sx=sx,my=my,sy=sy)$value)
    rho <- h.lloyd(y)/h.lloyd(x)
    H2 <- (3*J/(m*n*sqrt(2*pi)*(1+2*rho^2)^1.5*(2*rho^2*I4+I5)^2))^(1/7)
    H1 <- rho*H2

    return(list(H=H,H1=H1,H2=H2))
}

plugin <- function(x,y)
{
    bandsx <- pluginbands(x,y)
    bandsy <- pluginbands(y,x)
    cfg <- calc.cfg(bandsx$H,bandsx$H1,bandsx$H2,x,y,diag=1)
    cgf <- calc.cfg(bandsy$H,bandsy$H1,bandsy$H2,y,x,diag=1)
    m <- length(x)
    n <- length(y)    
    h1 <- cfg/(m^(1/3))
    h2 <- cgf/(n^(1/3))
    return(list(hx=h1,hy=h2))
}

calc.cfg <- function(H,H1,H2,x,y,diag=1)
{
    tmp <- crossval.int(H,H1,H2,x,y,diag=diag)
    cfg <- (0.5641896 *  tmp$int3 / abs(tmp$int2))^(1/3)
    return(cfg)
}



# Compute estimates for Delta function
crossval.int <- function(H,H1,H2,x,y,diag=1)
{
    m <- length(x)
    n <- length(y)
    x.K <- make.K(x,H1,y)
    y.K <- make.K(y,H,x)

    # Make f2 and g2, etc.
    f12 <- f12 <- glo <- numeric(n)
    g2 <- numeric(m)
    for(i in 1:m)
    {
        tmp <- matrix(y.K$K2[i,],nrow=n,ncol=1) %*% matrix(y.K$K2[i,],nrow=1,ncol=n)
        g2[i] <- 2*sum(tmp[lower.tri(tmp,diag=F)]) / (n*(n-1)*H^2)
    }
    for(i in 1:n)
    {
        tmp <- matrix(x.K$K3[i,],nrow=m,ncol=1) %*% matrix(x.K$K3[i,],nrow=1,ncol=m)
        if(diag==1)
            f12[i] <- (2*sum(tmp[lower.tri(tmp,diag=T)]) - sum(diag(tmp))) / (m*m*H1^4)
        else if(diag==0)
            f12[i] <- 2*sum(tmp[lower.tri(tmp,diag=F)]) / (m*(m-1)*H1^4)
        else
            f12[i] <- (2*sum(tmp[lower.tri(tmp,diag=T)]) - 1.5*sum(diag(tmp))) / (m*(m-0.5)*H1^4)
        glo[i] <- (sum(dnorm((y[i]-y)/H2)) - dnorm(0))/((n-1)*H2)
    }
    return(list(int1=mean(f12),int2=mean(f12*glo),int3=mean(g2),int4=mean(glo)))
}

## FUNCTIONS TO COMPUTE INTEGRALS CONTAINING NORMAL DENSITIES

fgderiv <- function(x,mx,sx,my,sy)
{
    xx <- (x-mx)/sx
    yy <- (x-my)/sy
    f <- dnorm(xx)/sx
    g <- dnorm(yy)/sy
    fdash <- -xx*f/sx
    gdash <- -yy*g/sy
    fdash2 <- (xx^2-1)*f/sx^2
    gdash2 <- (yy^2-1)*g/sy^2
    fdash3 <- xx*(3-xx^2)*f/sx^3
    gdash3 <- yy*(3-yy^2)*g/sy^3
    return(list(f=f,g=g,fdash=fdash,gdash=gdash,fdash2=fdash2,gdash2=gdash2,fdash3=fdash3,gdash3=gdash3))
}

## fg^2 assuming normal densities
fg2 <- function(x,mx,sx,my,sy)
{
    N <- fgderiv(x,mx,sx,my,sy)
    return(N$f*N$g^2)
}

## (f'g)^2 assuming normal densities
f1g2 <- function(x,mx,sx,my,sy)
{
    N <- fgderiv(x,mx,sx,my,sy)
    return((N$fdash*N$g)^2)
}

## f.f''.g
intf2 <- function(x,mx,sx,my,sy)
{
    N <- fgderiv(x,mx,sx,my,sy)
    return(N$f*N$fdash2*N$g)
}

## (fg)^2
intf3 <- function(x,mx,sx,my,sy)
{
    N <- fgderiv(x,mx,sx,my,sy)
    return((N$f*N$g)^2)
}


## g^2.f'.f'''
intf5 <- function(x,mx,sx,my,sy)
{
    N <- fgderiv(x,mx,sx,my,sy)
    return(N$fdash*N$fdash3*N$g^2)
}

## g g'' (f')^2
intf6 <- function(x,mx,sx,my,sy)
{
    N <- fgderiv(x,mx,sx,my,sy)
    return(N$fdash^2*N$gdash2*N$g)
}

## f^3.g
intf9 <- function(x,mx,sx,my,sy)
{
    N <- fgderiv(x,mx,sx,my,sy)
    return(N$f^3*N$g)
}

## g^3.f.(f')^2
intf10 <- function(x,mx,sx,my,sy)
{
    N <- fgderiv(x,mx,sx,my,sy)
    return(N$fdash^2*N$f*N$g^3)
}

## COMPUTE TRUE ROC FUNCTIONS


# Assuming normal densities
ROC.norm <- function(mx,sx,my,sy,nx=200)
{
    x <- seq(mx-5*sx,mx+5*sx,l=nx)
    y <- seq(my-5*sy,my+5*sy,l=nx)
    ROC(x,pnorm(x,mx,sx),y,pnorm(y,my,sy),nx=nx)
}

ROC.beta <- function(ax,bx,ay,by,nx=200)
{
    x <- seq(0,1,l=nx)
    ROC(x,pbeta(x,ax,bx),x,pbeta(x,ay,by),nx=nx)
}

ROC.t <- function(vx,vy,nx=200)
{
    x <- seq(-5,5,l=nx)
    ROC(x,pt(x,vx),x,pt(x,vy),nx=nx)
}

ROC.pareto <- function(cx,cy,ax,ay,nx=200)
{
    x <- seq(-5,5,l=nx)
    ROC(x,ppareto(x,cx,ax),x,ppareto(x,cy,ay),nx=nx)
}

ROC.gamma <- function(sx,sy,nx=200)
{
    x <- seq(0,10,l=nx)
    ROC(x,pgamma(x,sx),x,pgamma(x,sy),nx=nx)
}

ROC.t.mix <- function(mx1,mx2,vx1,vx2,my1,my2,vy1,vy2,nx=200,px1=0.5,px2=0.5,py1=0.5,py2=0.5)
{
    sump <- px1+px2
    px1 <- px1/sump
    px2 <- px2/sump
    sump <- py1+py2
    py1 <- py1/sump
    py2 <- py2/sump
    x <- seq(min(mx1,mx2)-9,max(mx1,mx2)+9,l=nx)
    y <- seq(min(my1,my2)-9,max(my1,my2)+9,l=nx)
    Fx <- px1*pt(x-mx1,vx1) + px2*pt(x-mx2,vx2)
    Fy <- py1*pt(y-my1,vy1) + py2*pt(y-my2,vy2)
    ROC(x,Fx,y,Fy,nx=nx)
}

# Assuming equally weighted normal mixtures
ROC.norm.mix <- function(mx1,mx2,sx1,sx2,my1,my2,sy1,sy2,nx=200,px1=0.5,px2=0.5,py1=0.5,py2=0.5)
{
    sump <- px1+px2
    px1 <- px1/sump
    px2 <- px2/sump
    sump <- py1+py2
    py1 <- py1/sump
    py2 <- py2/sump
    x <- seq(min(mx1,mx2)-5*max(sx1,sx2),max(mx1,mx2)+5*max(sx1,sx2),l=nx)
    y <- seq(min(my1,my2)-5*max(sy1,sy2),max(my1,my2)+5*max(sy1,sy2),l=nx)
    Fx <- px1*pnorm(x,mx1,sx1) + px2*pnorm(x,mx2,sx2)
    Fy <- py1*pnorm(y,my1,sy1) + py2*pnorm(y,my2,sy2)
    ROC(x,Fx,y,Fy,nx=nx)
}

## CALCULATION OF ASYMPTOTIC CONFIDENCE BANDS
## as per Hall, Hyndman and Fan (2003b)

ASMCONF.est <- function(x,y,alpha=0.05,ngrid=512,...)
{
    m <- length(x)
    n <- length(y)

    # Get h1 and h2
    h1 <- h.lloyd(x)/2
    h2 <- h.lloyd(y)/2

    # Get H1 and H2
    H1 <- (4/7/m)^{1/9}*sd(x)
    H2 <- (4/7/n)^{1/9}*sd(y)

    # Get Hf and Hg
    Hf <- bw.SJ(x, method="dpi")
    Hg <- bw.SJ(y, method="dpi")

    # Get HF and HG
    HF <- h.lloyd(x)
    HG <- h.lloyd(y)

    # Get hf and hg
    lo <- min(min(x)-5*Hf,min(y)-5*Hg)
    hi <- max(max(x)+5*Hf,max(y)+5*Hg)
    xx <- seq(lo,hi,l=ngrid)
    f <- pdf.est(x,Hf,xx)$y
    g <- pdf.est(y,Hg,xx)$y
    F <- cdf.est(x,HF,xx)$y
    G <- cdf.est(y,HG,xx)$y
    f2 <- denest(x,2,H1,xgrid=xx)$y
    g2 <- denest(y,2,H2,xgrid=xx)$y
    tt <- seq(0.001,0.999,l=200)
    Finv <- approx(unique(F),xx[!duplicated(F)],xout=1-tt)$y
    Ginv <- approx(unique(G),xx[!duplicated(G)],xout=1-tt)$y
    kappa <- 0.5641896
    z <- abs(qnorm(0.5*alpha))
    gFinv <- approx(xx,g,xout=Finv)$y
    fFinv <- approx(xx,f,xout=Finv)$y
    GFinv <- approx(xx,G,xout=Finv)$y
    f2Finv <- approx(xx,f2,xout=Finv)$y
    g2Finv <- approx(xx,g2,xout=Finv)$y
    tmp <- (n/m)*(gFinv/fFinv)^2*tt*(1-tt)
    a <- tmp/(GFinv*(1-GFinv)+tmp)
    qf <- kappa*(3-a-a*z^2)+2/sqrt(2*pi)*(a+a*z^2-1)
    qg <- kappa*(a-1-a*z^2)
    temp1 <- sum(qf[!is.na(f2Finv)])/sum(f2Finv,na.rm=TRUE)
    temp2 <- sum(qg[!is.na(f2Finv)])/sum(g2Finv,na.rm=TRUE)
    hf <- ifelse(temp1<0, 2^{-1/3}, 1) *(abs(temp1))^{1/3}*m^{-1/3}
    hg <- ifelse(temp2<0, 2^{-1/3}, 1) *(abs(temp2))^{1/3}*n^{-1/3}

    # Get ftilde and gtilde
    ft <- pdf.est(x,hf,xx)$y
    gt <- pdf.est(y,hg,xx)$y

    # Get Fhat and Ghat
    Fhat <- cdf.est(x,h1,xx)$y
    Ghat <- cdf.est(y,h2,xx)$y

    # Compute sigma^2
    Finv <- approx(unique(Fhat),xx[!duplicated(Fhat)],xout=1-tt)$y
    GFinv <- approx(xx,Ghat,xout=Finv)$y
    gFinv <- approx(xx,gt,xout=Finv)$y
    fFinv <- approx(xx,ft,xout=Finv)$y
    sig2 <- sqrt(1/n*GFinv*(1-GFinv) + 1/m*tt*(1-tt)*(gFinv/fFinv)^2)

    # Compute upper and lower bounds
    lo <- pmax(1-GFinv - z * sig2, 0)
    hi <- pmax(1-GFinv + z * sig2, 0)

    return(list(p=tt,Rp=1-GFinv, sigma2=sig2, lower=lo, upper=hi))
}

content/publications/witsml-funcs.R
# witsml-show50.R

library(xml2)
library(data.table)
library(dplyr)

as_named_list <- function(...) {
    if (length(as.character( match.call()[-1])) < 1)
        stop("you must supply at least one object name")
    nl <- setNames( list(...), as.character( match.call()[-1]) )
    nl
}


get_results_xml <- function() {
    witsml_dir <- "./witsml"
    all_files_xml <- list.files(witsml_dir, recursive = TRUE,
                                full.names = TRUE,
                                include.dirs = TRUE, pattern = "*.xml")
    
    # get the file for trajectory
    traj_files <- grep(pattern = "trajectory", ignore.case = TRUE,
                       value = TRUE, x = all_files_xml)
    
    dat <- read_xml(traj_files)
    
    # strip default namespaces from the document
    xml_ns_strip(dat)
    show_all_elems <- dat %>%
        xml_find_all( '//*') %>%
        xml_path()
    
    length_all_elems <- length(show_all_elems)
    
    # name of the orphan nodes
    orphan_vars <- c(1:14)
    vars14_names <- xml_name(xml_children(xml_find_first(dat, 
                                                         "//trajectory")))[orphan_vars]
    
    children_trajectory <- xml_name(xml_children(xml_find_all(dat, "//trajectory")))
    length_children_trajectory <- length(children_trajectory)
    
    # names of the nodes for "trajectoryStation"
    names_trajectoryStation <- xml_name(xml_children(xml_find_first( dat, "//trajectory/trajectoryStation")))
    
    # how many of //trajectoryStation/dTimStn
    trajectoryStation.dTimStn <- xml_find_all(dat, "//trajectoryStation/dTimStn")
    
    # we end up finding a way to calculate the number of trajectory stations
    number_of_trajectoryStation <- length(trajectoryStation.dTimStn)   
    
    # name of the dependent nodes of "trajectoryStation"
    children_trajectoryStation <- xml_name(xml_children(xml_find_first(dat, "//trajectoryStation")))
    
    return(as_named_list(
        traj_files, 
        show_all_elems,
        length_all_elems,
        vars14_names,
        children_trajectory,
        length_children_trajectory,
        names_trajectoryStation,
        number_of_trajectoryStation,
        children_trajectoryStation
        
    ))
}







static/Rfiles/funcfor.R
require(demography)

##########################
#### FRENCH MORTALITY ####
##########################

# Plot data
plot(fr.mort,series="male")

# Figure 1 : B&W version for paper with only a few years
plot(fr.mort,series="male",
    years=c(1899,1918,1957,1977,1997),type="p",cex=0.6,col=1,pch=c(1,15,22,2,16),
    main="France: Male death rates")
legend(5,1.5,legend=paste(c(1899,1918,1957,1977,1997)),col=1,pch=c(1,15,22,2,16),cex=0.8)

# Smoothing
france.sm <- smooth.demogdata(fr.mort,obs.var="theoretical")
i=c(1901,1973);plot(fr.mort,years=i,type="p",pch=1);lines(france.sm,years=i)

# Fit model to data up to 2001.
france.short <- extract.years(france.sm,1899:2001)
frmort.fit <- fdm(france.short,series="Male",order=4,lambda=3,max.age=110,method="M")
frmort.fcast <- forecast(frmort.fit,h=50,method="arima",level=80,adjust=FALSE)

# Figure 6: fitted basis functions and forecast coefficients
plot(frmort.fcast,"c",ylab1="B",ylab2="C",mean.lab="mux",main.title="",
    interaction.title="",shadecol=gray(0.9),fcol="black",outlier.col=1,outlier.pch=18)

# Forecasts with prediction intervals
years <- 2001+c(1,20)
plot(frmort.fcast,series="Male",years=years,ylim=c(-9.7,0),lty=1:2,col=1,max.age=109)
lines(frmort.fcast,series="upper",years=years,lty=1:2,col=1,max.age=109)
lines(frmort.fcast,series="lower",years=years,lty=1:2,col=1,max.age=109)
legend(0,0,legend=paste(years),lty=1:2,cex=0.7)

plot(frmort.fcast,series="Male",years=years,lty=1:2,col=1,max.age=100,transform=FALSE)
lines(frmort.fcast,series="upper",years=years,transform=FALSE,lty=1:2,col=1,max.age=100)
lines(frmort.fcast,series="lower",years=years,transform=FALSE,lty=1:2,col=1,max.age=100)
legend(0,0.43,legend=paste(years),lty=1:2,cex=0.7)
rm(years)


##############################
#### AUSTRALIAN FERTILITY ####
##############################

# Use data up to 2000
aus.fert$lambda <- 0 # Use logs for transformation
ausfert <- extract.years(aus.fert,1921:2000)
ausfert$rate$female <- rbind(rep(0.005,80),ausfert$rate$female,rep(0.005,80))
ausfert$pop$female <- rbind(ausfert$pop$female[1,],ausfert$pop$female,ausfert$pop$female[7,])
rownames(ausfert$rate$female) <- rownames(ausfert$pop$female) <- c("<15","15-19","20-24","25-29","30-34","35-39","40-44","45-49",">49")
ausfert$age <- c(13,17,22,27,32,37,42,47,52)

# Figure 2: plot separate time series for each age group
plot(aus.fert,plot.type="time",xlim=c(1920,2005),lty=1:7,show.labels=FALSE,transform=TRUE)
text(rep(2004,7)-c(0,0,8,8,0,0,0),log(aus.fert$rate[[1]])[,80]+c(0,.1,.4,-.2,-.1,0,0),
    rownames(aus.fert$rate$female),cex=0.7)

# Figure 2. published version
plot(aus.fert,years=1921:2000,plot.type="time",xlim=c(1920,2005),lty=1:7,show.labels=FALSE,col=1,transform=TRUE,main="",xlab="Year")
text(rep(2004,7)-c(0,0,8,8,0,0,0),log(aus.fert$rate[[1]])[,80]+c(0,.1,.4,-.2,-.1,0,0),
    rownames(aus.fert$rate$female),cex=0.7)

# Smooth data
fert.sm <- smooth.demogdata(ausfert,age.grid=13:52,obs.var="theoretical")
plot(fert.sm,transform=TRUE);points(ausfert,transform=TRUE)

# Plot of smoothed data
years=c(1921,1940,1960,1980,2000)
plot(fert.sm,years=years,col=1:5,main="")
points(ausfert,years=years,pch=1,col=1:5,cex=0.6)
legend("topright",legend=paste(years),lty=1,col=1:5,cex=0.9)

plot(fert.sm,years=years,col=1:5,main="",transform=TRUE)
points(ausfert,years=years,pch=1,col=1:5,cex=0.6,transform=TRUE)
legend("topright",legend=paste(years),lty=1,col=1:5,cex=0.9)

# Figure 3:
plot(fert.sm,years=years,lty=1:5,col=1,main="",transform=TRUE)
points(ausfert,years=years,pch=1,col=1,cex=0.6,transform=TRUE)
legend(20,-1,legend=paste(years),lty=1:5,cex=0.9)
rm(years)

# Fit model and forecasting using ARIMA
ausfert.fit <- fdm(fert.sm,order=3,method="classical")
fert.fcast <- forecast(ausfert.fit,h=20,level=80,method="arima",adjust=FALSE)

# Figure 4: Plot of forecast coefficients
plot(fert.fcast,plot.type="c",ylab2="C",ylab1="B",mean.lab="mux",main.title="",interaction.title="",shadecols=gray(0.9),fcol=1)

# Figure 5: Plot of forecast curves
years=2000+c(1,20)
plot(fert.fcast,years=years,lty=1:2,col=1,ylim=c(0,160),yaxs="i",main="")
lines(fert.fcast,series="upper",years=years,lty=1:2,col=1)
lines(fert.fcast,series="lower",years=years,lty=1:2,col=1)
legend(15,145,legend=paste(years),lty=1:2,cex=0.7)
rm(years)
